conda-store-server/conda_store_server/action/generate_constructor_installer.py

```diff
@@ -0,0 +1,108 @@
+import os
+import pathlib
+import subprocess
+import sys
+import tempfile
+
+import yaml
+from conda_store_server import action, schema
+
+
+@action.action
+def action_generate_constructor_installer(
+    context,
+    conda_command: str,
+    specification: schema.CondaSpecification,
+    installer_dir: pathlib.Path,
+    version: str,
+):
+    # Helpers
+    def print_cmd(cmd):
+        context.log.info(f"Running command: {' '.join(cmd)}")
+        context.log.info(
+            subprocess.check_output(cmd, stderr=subprocess.STDOUT, encoding="utf-8")
+        )
+
+    def write_file(filename, s):
+        with open(filename, "w") as f:
+            context.log.info(f"{filename}:\n{s}")
+            f.write(s)
+
+    # pip dependencies are not directly supported by constructor, they will be
+    # installed via the post_install script:
+    # https://github.com/conda/constructor/issues/515
+    # conda and pip need to be in dependencies for the post_install script
+    dependencies = ["conda", "pip"]
+    pip_dependencies = []
+    for d in specification.dependencies:
+        if type(d) is schema.CondaSpecificationPip:
+            pip_dependencies.extend(d.pip)
+        else:
+            dependencies.append(d)
+
+    # Creates the construct.yaml file and post_install script
+    ext = ".exe" if sys.platform == "win32" else ".sh"
+    pi_ext = ".bat" if sys.platform == "win32" else ".sh"
+    installer_filename = (installer_dir / specification.name).with_suffix(ext)
+
+    os.makedirs(installer_dir, exist_ok=True)
+
+    with tempfile.TemporaryDirectory(ignore_cleanup_errors=True) as tmp_dir:
+        tmp_dir = pathlib.Path(tmp_dir)
+        construct_file = tmp_dir / "construct.yaml"
+        post_install_file = (tmp_dir / "post_install").with_suffix(pi_ext)
+        env_dir = tmp_dir / "env"
+
+        construct = {
+            "installer_filename": str(installer_filename),
+            "post_install": str(post_install_file),
+            "name": specification.name,
+            "channels": specification.channels,
+            "specs": dependencies,
+            "version": version,
+        }
+
+        if sys.platform == "win32":
+            post_install = """\
+call "%PREFIX%\Scripts\activate.bat"
+"""
+        else:
+            post_install = """\
+#!/usr/bin/env bash
+set -euxo pipefail
+source "$PREFIX/etc/profile.d/conda.sh"
+conda activate "$PREFIX"
+"""
+        if pip_dependencies:
+            post_install += f"""
+pip install {' '.join(pip_dependencies)}
+"""
+
+        # Writes files to disk
+        write_file(construct_file, yaml.dump(construct))
+        write_file(post_install_file, post_install)
+
+        # Installs constructor
+        command = [
+            conda_command,
+            "create",
+            "-y",
+            "-p",
+            str(env_dir),
+            "constructor",
+        ]
+        print_cmd(command)
```

Comment:
It _is_ wasteful to download `constructor` every time, but I do it for the following reasons:
- I don't want to assume that `constructor` is already available in the user environment, so I need to install it
- I'm not sure I can rely on conda-store installation method to provide runtime deps. E.g., if users install conda-store via `pip`, how would they get `constructor`?
- This needs to be installed into a separate env to avoid clashing with user packages in standalone mode
- This ensures the latest `constructor` version is used, without requiring manual intervention
- If a failure occurs for whatever reason, the next call to this action will try again (actions are state-independent).

I think we can try the following in a _separate PR_, which I didn't implement due to time constraints:
- have a predefined env name, which is unlikely to clash with user envs, like `conda-store-runtime`, and installing `constructor` there on first call of this action
- then checking if it's already available in the env by calling `constructor --help`
- if a new version is available upstream, downloading and installing that one
- if `--help` fails, I think the entire environment needs to be removed because it might be in an invalid state (a user hit Ctrl-C while this action was running, network error, etc.)
- removing the env needs to be done via python APIs for dir removal and not via `conda env remove` because I had issues with it where it failed to remove an existing env.

Reply:
Actually, we might not want to update automatically just to avoid bumping into unexpected issues. Instead, we could pin the constructor version here and update it ourselves once in a while if new features are needed.

Reply:
I really don't like this. We should add it as a runtime dependency. If `pip install` compatibility is a problem, it is too for the `conda` dependency. Worst case we can try-import it and disable the option if `constructor` is not available. It will save tons of complexity we don't really need.

Reply:
> Worst case we can try-import it and disable the option if constructor is not available.

I like this suggestion! This solves the compatibility/distribution problem while still making this feature available.